sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/bigquery/BigQueryTable.java

```diff
@@ -39,13 +45,36 @@
 class BigQueryTable extends BaseBeamTable implements Serializable {
   @VisibleForTesting final String bqLocation;
   private final ConversionOptions conversionOptions;
+  private Double tableSize = -1.;
 
   BigQueryTable(Table table, BigQueryUtils.ConversionOptions options) {
     super(table.getSchema());
     this.conversionOptions = options;
     this.bqLocation = table.getLocation();
   }
 
+  @Override
+  public Double getRowCount(PipelineOptions options) {
+    if (tableSize == -1.) {
```

Comment:
This is also a magic constant that means "table size value has not been set". This is hard to realized without reading the whole code. The reason is because with `Double` you don't have a distinction between "value not set", "unknown", or any other value that is not an actual row count. If you had a custom class, e.g. `BeamTableStatistic`, you could have a default value there similar to `Statistic.UNKNOWN`. This way this code would look like:

```
if (statistic == null) {
  statistic = getRowCountFromBQ(...)
}
return statistic;
```

and would help readability a bit if you move the logic into a separate method:

```
private static BeamTableStatistic getRowCountFromBQ(PipelineOptions o, String bqLocation) {
   try {
      BigInteger rowCount = BigQueryHelpers.getNumRows(...);
      return BeamTableStatistic.of(rowCount);
   }  catch (...) {
      return BeamTableStatistic.unknown();
   }
}
```